core/src/main/java/org/apache/calcite/jdbc/SimpleCalciteSchema.java

```diff
@@ -67,33 +69,72 @@ public CalciteSchema add(String name, Schema schema) {
     return calciteSchema;
   }
 
+  private String caseInsensitiveLookup(Set<String> candidates, String name) {
+    // Exact string lookup
+    if (candidates.contains(name)) {
+      return name;
+    }
+    // Upper case string lookup
+    final String upperCaseName = name.toUpperCase(Locale.ROOT);
+    if (candidates.contains(upperCaseName)) {
+      return upperCaseName;
+    }
+    // Lower case string lookup
+    final String lowerCaseName = name.toLowerCase(Locale.ROOT);
+    if (candidates.contains(lowerCaseName)) {
+      return lowerCaseName;
+    }
+    // Fall through: Set iteration
+    for (String candidate: candidates) {
+      if (candidate.equalsIgnoreCase(name)) {
+        return candidate;
+      }
+    }
+    return null;
+  }
+
   protected CalciteSchema getImplicitSubSchema(String schemaName,
       boolean caseSensitive) {
     // Check implicit schemas.
-    Schema s = schema.getSubSchema(schemaName);
-    if (s != null) {
-      return new SimpleCalciteSchema(this, s, schemaName);
+    final String schemaName2 = caseSensitive ? schemaName : caseInsensitiveLookup(
+        schema.getSubSchemaNames(), schemaName);
+    if (schemaName2 == null) {
+      return null;
     }
-    return null;
+    final Schema s = schema.getSubSchema(schemaName2);
+    if (s == null) {
+      return null;
+    }
+    return new SimpleCalciteSchema(this, s, schemaName2);
   }
 
   protected TableEntry getImplicitTable(String tableName,
       boolean caseSensitive) {
     // Check implicit tables.
-    Table table = schema.getTable(tableName);
-    if (table != null) {
-      return tableEntry(tableName, table);
+    final String tableName2 = caseSensitive ? tableName : caseInsensitiveLookup(
+        schema.getTableNames(), tableName);
+    if (tableName2 == null) {
```

Comment:
Can we abstract a matcher like `SqlNameMatcher` for name matching ?

Reply:
I ever thought about abstracting an interface. But it may bring complexity.
(1) Different with name matcher, case (in)sensitive lookup is not widely used. Even for `CalciteSchema`, there is already a more general and efficient interface `NamedMap`. This piece of code is just for `SimpleCalciteSchema`, without maintaining cache.
(2) For performance, we should prevent to call `getTableNames`/`getSubSchemas`/`getTypeNames` when case sensitive. Therefore, if we abstract the logic, the interface arguments need to contain functions. It may be too complex. 
Do you have any suggestions?

Reply:
I thought if we can bookeep the member names for duplicate, one is with key that same with the original name, one is with a UPPERCASE name, just like `ListSqlOperatorTable` did.

Reply:
That may be same as `CachingCalciteSchema`. I guess if we bookeep or maintain something (e.g., `CaseSensitiveKey`, `CaseInSensitiveKey`), the structure will be optimized and evolved to the efficient `NameSet`, as that in `CachingCalciteSchema` (i.e., `Cache<NameSet>`).

`CachingCalciteSchema` is used by default in Calcite. According to their names, `SimpleCalciteSchema` may be different, and query items (i.e., schema/table/type) on the fly?

Reply:
In Flink, we did use the `SimpleCalciteSchema` directly, and we do not have any requests for the case-sensitivity now, out UDT and UDF all not stored in this schema, but maintained by Flink itself.

I would rather to keep the `SimpleCalciteSchema` as simple as possible, can we just keep it as it is now and do not support the case-sensitivity look-up ? For them to use the `CachingCalciteSchema ` if they what the case-insensitive look up.

Reply:
>  can we just keep it as it is now and do not support the case-sensitivity look-up

 Case-insensitive? Current SimpleCalciteSchema supports only case-sensitive lookup.

>  keep the SimpleCalciteSchema as simple as possible

 Do you mean the simple code style? In logic, this PR adds a branch for case-insensitive, bringing no burden to case-sensitive branch.
 
In our production environment, we use the default `CachingCalciteSchema`. Therefore, I do not have strong requirement for this feature. The only worry is that users may get incorrect result when using `SimpleCalciteSchema` for case-insensitive lookup.
Since this Jira issue is filed by @zabetak, how about your opinion?

Reply:
I see two options:
1. We provide a way to do case-insensitive lookups without affecting the performance of the case-sensitive case (so exactly what this PR tries to do);
2. We forbid case-insensitive lookups by throwing an `IllegalArgumentException` when the flag is `true`. 
Leaving as it is now right now is confusing and might cause subtle bugs that may not reveal themselves from the beginning. I am OK with any of those two options.

Reply:
@DonnyZone Moving discussion here. why open a new PR? what's wrong with this one?

Reply:
> @DonnyZone Moving discussion here. why open a new PR? what's wrong with this one?

There are conflicts now. I will take time to rebase it.